cmd/sourced/compose/file/file.go

```diff
@@ -98,7 +102,13 @@ func InitDefaultOverride() (string, error) {
 
 // Download downloads the docker-compose.yml file from the given revision
 // or URL. The file is set as the active compose file.
-func Download(revOrURL RevOrURL) error {
+func Download(revOrURL RevOrURL) (err error) {
+	defer func() {
```

Comment:
I don't know how much I like the usage of `defer` to just wrap an error. Is this an idiomatic go thing? I mean I'd find the usefulness of this if you want to also recover from panic and maybe also wrap the panic error. But in this case, maybe a simple wrapper function would do the trick?

Reply:
do you mean something like the following?

```go
func Download(revOrURL RevOrURL) error {
  err := downloadWrappedErr(revOrURL)
  if err != nil {
    return ErrCouldNotGetComposer.Wrap(err)
  }

  return nil
}

func downloadWrappedErr(revOrURL RevOrURL) error {
  // original content of Download(RevOrURL) error
}
```

It seems ugly imo

On the other hand, if I understood the proposal for `try`, handling and wrapping errors in `defer` seems to be something proposed by go team, isn't it?
>_We advocate using the existing defer statement and standard library functions to help with augmenting or wrapping of errors._

https://github.com/golang/go/issues/32437

Does not seems a clean and explicit way to handle and wrap these errors?

Reply:
yup, but that proposal has been closed and rejected due to a lot of criticism AFAIU.

Reply:
BTW I was thinking something as follows:
```
func Download(revOrURL RevOrURL) error {
  wf := func(err) {
    if err != nil {
      return ErrCouldNotGetComposer.Wrap(err)
    }
    return nil
  }

  ...
    return wf(err)
  ...

  return nil
}
```

Reply:
But I'm not a golang expert, so I really cannot strongly advocate some styles over others.

Reply:
You don't need separate function wrapper. You can wrap it in the caller.
For example `getOrInstallComposeContainer` function returns normal errors and `newCompose` function return `ErrCouldNotGetComposer`.

If `getOrInstallComposeContainer` is called from another place it already knows that `get` has failed as as it just called it. But it may want to inspect underlying errors and return another type of high-level error. A good example of such nesting described by Russ Cox in https://blog.golang.org/experiment

Reply:
The goal of `defer` is to ensure that `Download` returns a `ErrConfigDownload`, no matter where it's called.

If not used the `defer`, I see three alternatives:
- **a)** make sure we wrap the returned error wherever `Download` is called, as I understood from @smacker https://github.com/src-d/sourced-ce/pull/247#discussion_r326152372,
  ```go
  // in cmd/sourced/compose/file/file.go
  func InitDefault() (string, error) {
  	...
  	err := Download(version)
  	if err != nil {
  		return "", ErrConfigDownload.Wrap(err)
  	}
  	return activeFilePath, nil
  }

  // in cmd
  func (c *composeDownloadCmd) Execute(args []string) error {
  	...
  	err := Download(version)
  	if err != nil {
  		return ErrConfigDownload.Wrap(err)
  	}
  	return nil
  }
  ```
  what is error-prone imo, because `Download` can be called elsewhere tomorrow, so then it could be not wrapped.
- **b)** make sure that we wrap all the returned errors inside `Download`, what is repetitive and can be also error-prone if a new return is added later and not wrapped. That's what I understood from @se7entyse7en's https://github.com/src-d/sourced-ce/pull/247#discussion_r326145898 
- **c)** do as suggested by https://github.com/src-d/sourced-ce/pull/247#discussion_r326140663, moving the current content of `Download` into a new (and unexported) function which will do the actual work, and then call it from (the exported) `Download`, and wrap and return its error.
This would be bulletproof and not error-prone (if the new unexported function is not directly called from the same package elsewhere), but as I said, I find this procedure ugly and could be an anti-pattern.

If you strongly oppose to deferring the error handling, and you don't suggest another option, I'd vote for **b** as @se7entyse7en suggested, because even being error-prone as I said, the risk is limited to the scope of the same function.

Reply:
I also like option **c)** actually, and I don't think it's anti-pattern. For example have a look at code for `net/http` in `client.go` [here](https://github.com/golang/go/blob/master/src/net/http/client.go#L508). It uses an expored method that only calls an unexported one. In this case there's no error handling, but I don't think would be considered an anti-pattern.

Reply:
> what is error-prone imo, because Download can be called elsewhere tomorrow, so then it could be not wrapped.

you assume that caller from the different location must behave exactly like the current one, which isn't true.
make-up example:
```go
err := Download(url)
// with current code I have to know that `Download` returns `go-error` type
// which I must inspect using `Case()` to make this check
if err == NetworkError {
  return LocalFallbackUpdate()
}
// with current code produces weird text like
// "can't set new config: can't download config: no permission to change active config"
return ErrCanNotSetNewConfig.Wrap(err)
```

As you can see the error is just incorrect. It _could_ download the config. It failed on `SetActive`.

Reply:
@se7entyse7en jfyi we discussed it offline and I found a better explanation which David agreed with. With current state of go 1.12 you usually return errors "one level less". Let me give you an example:

```go
func downloadFile() error {
  // unwrapped
  return err
}

func setComposeFIle() error {
  // unwrapped
  return err
}

func UpdateCompose() error {
  if err := downloadFile(); err != nil {
    return ErrDownload
  }

  if err := setComposeFIle(); err != nil {
    return ErrSetComposeFile
  }
}
```

You don't need to wrap an error in the function `downloadFile` as the caller already knows it is download error. That information comes from the name of the function and `UpdateCompose` function can inspect if it was network or fs error and do something else (like fallback) using only current features of the language without pulling external lib like `go-error` into the codebase. On another hand `UpdateCompose` (if it makes sense) wraps error underlying errors to allow the caller distinguish what happened. (without wrap you don't know if fs error was caused by downloading of set)

Reply:
Thanks for the explanation! Seems really clear!